File: src/main/java/org/noxet/noxetserver/creepersweeper/CreeperSweeperGame.java
package org.noxet.noxetserver.creepersweeper;

import org.noxet.noxetserver.util.InventoryCoordinate;
import org.noxet.noxetserver.util.InventoryCoordinateUtil;

import java.util.*;

public class CreeperSweeperGame {
    public enum CreeperSweeperStatus {
        NOT_STARTED,
        PLAYING,
        ENDED
    }

    private final int width, height, creepers;
    private final Map<InventoryCoordinate, CreeperSweeperTile> tiles = new HashMap<>();
    private CreeperSweeperStatus status = CreeperSweeperStatus.NOT_STARTED;
    private boolean hasWon = false;
    private long gameStart;

    public CreeperSweeperGame(int width, int height, int creepers) {
        this.width = width;
        this.height = height;
        this.creepers = creepers;

        for(int i = 0; i < width * height; i++)
            tiles.put(InventoryCoordinateUtil.getCoordinateFromSlotIndex(i), new CreeperSweeperTile()); // Placeholder tiles.
    }

    public void start(InventoryCoordinate startCoordinate) {
        if(status != CreeperSweeperStatus.NOT_STARTED)
            return;

        status = CreeperSweeperStatus.PLAYING;

        tiles.clear(); // Clear placeholder tiles.

        List<CreeperSweeperTile> nonCreeperTiles = new ArrayList<>();

        for(int i = 0; i < width * height; i++) {
            CreeperSweeperTile tile = new CreeperSweeperTile();
            InventoryCoordinate tileCoordinate = InventoryCoordinateUtil.getCoordinateFromSlotIndex(i);
            tiles.put(tileCoordinate, tile);
            if(!tileCoordinate.isAt(startCoordinate))
                nonCreeperTiles.add(tile);
        }

        assert creepers <= width * height - 1;
        Random random = new Random();

        int creepersLeft = creepers;

        while(creepersLeft > 0) {
            nonCreeperTiles.remove(random.nextInt(nonCreeperTiles.size())).makeCreeperTile();
            creepersLeft--;
        }

        gameStart = System.currentTimeMillis();
    }

    public void stop() {
        if(status == CreeperSweeperStatus.ENDED)
            return;

        status = CreeperSweeperStatus.ENDED;
    }

    public long getGameDuration() {
        return System.currentTimeMillis() - gameStart;
    }

    public CreeperSweeperTile getTileAt(InventoryCoordinate coordinate) {
        Iterator<Map.Entry<InventoryCoordinate, CreeperSweeperTile>> tileEntries = tileIterator();

        while(tileEntries.hasNext()) {
            Map.Entry<InventoryCoordinate, CreeperSweeperTile> tileEntry = tileEntries.next();
            if(tileEntry.getKey().isAt(coordinate))
                return tileEntry.getValue();
        }

        return null;
    }

    public Iterator<Map.Entry<InventoryCoordinate, CreeperSweeperTile>> tileIterator() {
        return tiles.entrySet().iterator();
    }

    public boolean isPlaying() {
        return status == CreeperSweeperStatus.PLAYING;
    }

    public boolean hasEnded() {
        return status == CreeperSweeperStatus.ENDED;
    }

    public boolean didWin() {
        return hasWon;
    }

    /**
     * Performs a recursive tile reveal.
     * @param coordinate The tile to reveal
     */
    public void revealTileAt(InventoryCoordinate coordinate) {
        if(!isPlaying())
            start(coordinate);

        CreeperSweeperTile tile = getTileAt(coordinate);

        if(tile.isFlagged())
            return; // Cannot reveal flagged tile!

        if(tile.isCreeperTile()) {
            onRevealCreeperTile(coordinate);
            return;
        }

        tile.revealRecursively(this, coordinate, new ArrayList<>());

        Iterator<Map.Entry<InventoryCoordinate, CreeperSweeperTile>> tileEntries = tileIterator();

        while(tileEntries.hasNext()) {
            Map.Entry<InventoryCoordinate, CreeperSweeperTile> tileEntry = tileEntries.next();
            CreeperSweeperTile checkTile = tileEntry.getValue();
            if(!checkTile.isCreeperTile() && !checkTile.isRevealed())
                return;
        }

        onFinish();
    }

    public void onRevealCreeperTile(InventoryCoordinate coordinate) {
        stop();
    }

    public void onFinish() {
        hasWon = true;
        stop();
    }
}

File: src/main/java/org/noxet/noxetserver/creepersweeper/CreeperSweeperGame.scala
package org.noxet.noxetserver.creepersweeper

import org.noxet.noxetserver.util.InventoryCoordinate

import scala.collection.immutable.HashMap
import scala.collection.mutable
import scala.util.Random

class CreeperSweeperGame(width: Int, height: Int, creepers: Int):
  private enum CreeperSweeperStatus:
    case NotStarted, Playing, Ended

  val tiles: HashMap[InventoryCoordinate, CreeperSweeperTile] =
    (0 until width * height).map(i => (InventoryCoordinate.fromSlotIndex(i), CreeperSweeperTile()))
  private var status = CreeperSweeperStatus.NotStarted
  private var hasWon = false
  private var gameStart: Option[Long] = None

  def start(startCoordinate: InventoryCoordinate): Unit =
    if status != CreeperSweeperStatus.NotStarted then return

    status = CreeperSweeperStatus.Playing

    assert(creepers <= width * height - 1)
    tiles.filterNot(x => x._1 == startCoordinate)
      .sampleMany(creepers)(Random()).get.foreach(_.creeper = true)

    gameStart = Some(System.currentTimeMillis)

  def stop(): Unit = status = CreeperSweeperStatus.Ended

  def getGameDuration: Long = System.currentTimeMillis - gameStart.getOrElse(0L)

  def isPlaying: Boolean = status == CreeperSweeperStatus.Playing

  def hasEnded: Boolean = status == CreeperSweeperStatus.Ended

  def didWin: Boolean = hasWon

  /**
   * Performs a recursive tile reveal.
   *
   * @param coordinate The tile to reveal
   */
  def revealTile(coordinate: InventoryCoordinate): Unit =
    if !isPlaying then
      start(coordinate)

    val tile = tiles(coordinate)

    if tile.flagged then return // Cannot reveal flagged tile!

    if tile.creeper then
      onRevealCreeperTile(coordinate)
      return

    tile.revealRecursively(this, coordinate, List())

    if !tiles.exists(x => !x._2.creeper && !x._2.revealed) then
      // If all non-creeper tiles are revealed.
      onFinish()

  def onRevealCreeperTile(coordinate: InventoryCoordinate): Unit = stop()

  def onFinish(): Unit =
    hasWon = true
    stop()

  def getNeighborTiles(coord: InventoryCoordinate): HashMap[InventoryCoordinate, CreeperSweeperTile] = {
    for
      dx <- 0 until 3
      dy <- 0 until 3
      if !(dx == 1 && dy == 1)
    yield
      val x = coord.x - 1 + dx
      val y = coord.y - 1 + dy
      if !(x >= 9 || x < 0) then // Only X is modulus-ed so we don't need to check Y for out of bounds.
        val neighborCoord = InventoryCoordinate(x, y)
        tiles.get(neighborCoord).map((neighborCoord, _))
  }.toSet.flatMap

  def countCreeperNeighbors(coordinate: InventoryCoordinate): Int =
    getNeighborTiles(coordinate).count(_._2.creeper)

  /**
   * Reveals a tile, and neighbor tiles, recursively.
   */
  def revealRecursively(coordinate: InventoryCoordinate, ignore: mutable.HashSet[InventoryCoordinate]): Unit =
    tiles(coordinate).revealed = true

    if countCreeperNeighbors(coordinate) != 0 then return

    ignore.add(coordinate)

    getNeighborTiles(coordinate)
      .filterNot(x => ignore.contains(x._1))
      .foreach(x => revealRecursively(x._1, ignore))